packages/dds/tree/README.md

```diff
@@ -232,3 +232,85 @@ graph LR;
     ForestIndex--"invalidates"-->app
     Indexes--"delta (for apps that want deltas)"-->app
 ```
+
+### Dependencies
+
+`@fluid-internal/tree` depends on the Fluid runtime (various packages in `@fluidframework/*`)
+and will be depended on directly by application using it (though at that time it will be moved out of `@fluid-internal`).
+`@fluid-internal/tree` is also complex,
+so its implementation is broken up into several parts with have carefully controlled dependencies to help ensure the codebase is maintainable.
+The goal of this internal structuring is to make evolution and maintenance easy.
+Some of the principals used to guide this are:
+
+- Avoid cyclic dependencies:
+
+    Cyclic dependencies can make it hard to learn a codebase incrementally, as well as make it hard to update or replace parts of the codebase incrementally.
+    Additionally they can cause runtime issues with initialization.
+
+- Minimize coupling:
+
+    Reducing the number and complexity of edges in the dependency graph.
+    This often involves approaches like making a component generic instead of depending on a concrete type directly,
+    or combining related components that have a lot of coupling.
+
+- Reducing transitive dependencies:
+
+    Try to keep the total number of dependencies of a given component small when possible, with a particular emphasis on avoiding stateful dependencies for code with complex conditional logic.
```

Comment:
It's not completely clear to me what is meant by stateful dependencies.

Reply:
I rewrote and extended this section. Hopefully its more clear now